File: aoc2021kotlin/src/main/kotlin/Puzzle04.kt
package net.blergh.advent2021

import java.io.File

data class Tile(
    val bingoNum: Int,
    var marked: Boolean,
)

class Board(
    val boardId: Int,
    private val fiveInputLines: List<String>
)
{
    // name is a reminder that the indexing is e.g. foo[y][x] / foo[row][col]
    private val yxGrid: List<List<Tile>> = fiveInputLines.map { line ->
        line.trim().split("\\s+".toRegex()).map { bingoNumber ->
            Tile(Integer.parseInt(bingoNumber, 10), false)
        }
    }

    private var isWinner = false


    private fun isWinnerInRow(rowNum: Int): Boolean {
        val rowOfTiles = yxGrid[rowNum]
        for(tile in rowOfTiles) {
            if(!tile.marked) {
                return false
            }
        }
        return true
    }

    private fun isWinnerInColumn(colNum: Int): Boolean {
        val colOfTiles = yxGrid.map { row ->
            row[colNum]
        }
        for(tile in colOfTiles) {
            if(!tile.marked) {
                return false
            }
        }
        return true
    }


    fun markAndCheckForWin(bingoNum: Int): Boolean {
        for(colNum in 0..4) {
            for(rowNum in 0..4) {
                if(bingoNum == yxGrid[rowNum][colNum].bingoNum) {
                    yxGrid[rowNum][colNum].marked = true
                    isWinner = isWinnerInRow(rowNum) or isWinnerInColumn(colNum)
                    return isWinner
                }
            }
        }
        return false
    }

    fun isWinner(): Boolean {
        return isWinner
    }

    fun getScore(winningNumber: Int): Int {
        if(!isWinner) {
            return 0
        }

        return yxGrid.flatMap { row ->
            row.map { tile ->
                if(!tile.marked) tile.bingoNum else 0
            }
        }.sum() * winningNumber
    }

    override
    fun toString(): String {
        val name = "boardId=$boardId"
        val gridString = yxGrid.map { row ->
            row.map { tile ->
                val tf = if(tile.marked) "t" else "f"
                "${tile.bingoNum}${tf}"
            }.joinToString(separator = ",")
        }.joinToString(separator = "\n")

        return "$name\n$gridString"
    }
}

object Puzzle04 {
    fun run() {
        val input04 = File("${Main.aocRoot}/other/04/input04").readText().trim()
        val input04Chunks = input04.split("\n\n")

        val bingoSequence: List<Int> = input04Chunks[0].split(",").map { Integer.parseInt(it, 10) }
        val boards: List<Board> = input04Chunks.drop(1).mapIndexed { i, chunk ->
            val lines = chunk.split("\n")
            Board(i, lines)
        }

        var boardsStillInPlay = boards.size

        for (bingoNum in bingoSequence) {
            println("marking $bingoNum...")

            val remainingBoards = boards.filter { !it.isWinner() }
            for (board in remainingBoards) {
                if (board.markAndCheckForWin(bingoNum)) {
                    boardsStillInPlay -= 1

                    println("winner:\n$board")
                    println("$boardsStillInPlay boards remain")

                    //part 1
                    // To guarantee victory against the giant squid, figure out which board will win first.
                    // What will your final score be if you choose that board?
                    if (boardsStillInPlay == boards.size - 1) {
                        val score = board.getScore(winningNumber = bingoNum)
                        println("that was the FIRST winning board!")
                        println("(p1 answer) score = $score") // 23177
                    }

                    //part 2
                    // Figure out which board will win last.
                    // Once it wins, what would its final score be?
                    if (boardsStillInPlay == 0) {
                        val score = board.getScore(winningNumber = bingoNum)
                        println("that was the LAST winning board!")
                        println("(p2 answer) score = $score") // 6804
                    }
                }
            }
        }

        println("all numbers played")
    }
}

File: other/04/puzzle04.py
import re


def mark(board, number) -> None:
    for x in range(0, 5):
        for y in range(0, 5):
            if board[(x,y)][0] == number:
                board[(x,y)] = (number, True)


def is_winner(board) -> bool:
    # test rows
    for y in range(0, 5):
        row_marks = [board[(x,y)][1] for x in range(0, 5)]
        if row_marks == [True]*5:
            return True

    # test columns
    for x in range(0, 5):
        col_marks = [board[(x,y)][1] for y in range(0,5)]
        if col_marks == [True]*5:
            return True

    return False


def get_score(board, winning_number) -> int:
    total = sum([x[0] for x in board.values() if not x[1]])
    return total * winning_number


if __name__ == "__main__":
    input04 = open("input04", encoding="utf-8").read().strip()
    input04_chunks = input04.split("\n\n")

    bingo_sequence = [int(x) for x in input04_chunks[0].split(",")]

    board_list = []
    for chunk in input04_chunks[1:]:
        numbers = [int(x) for x in re.split("\\s+", chunk.strip())]
        board = {}
        for i in range(0,25):
            y = i // 5
            x = i % 5
            board[(x,y)] = (numbers[i], False) # tuple is (number, marked)
        board_list.append(board)


    for bingo_num in bingo_sequence:
        print(f"marking {bingo_num}...")

        boards_remaining = list(filter(lambda b: not is_winner(b), board_list))
        for board in boards_remaining:
            mark(board, bingo_num)
            if is_winner(board) and len(boards_remaining) == len(board_list):
                score = get_score(board=board, winning_number=bingo_num)
                print("FIRST winning board found!")
                print(board)
                print(f"(p1 answer) score = {score}") # 23177
            if is_winner(board) and len(boards_remaining) == 1:
                score = get_score(board=board, winning_number=bingo_num)
                print("LAST winning board found!")
                print(board)
                print(f"(p2 answer) score = {score}") # 6804

    print("bingo numbers exhausted")
